### PR TITLE
New module for managing volumes in Vexata storage arrays. (redo of PR #47091)

### PR DESCRIPTION
**Pull request by vexata**
Initial commit for module to manage Vexata storage volumes + fixes from code review in pr #47091 that got messed up.

##### SUMMARY
Adding new module for managing volumes in Vexata storage arrays. + fixes from code review in pr #47091 that got messed up.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vexata_volume

##### ANSIBLE VERSION
ansible 2.8.0.dev0 (vexata_volume2 962946954f) last updated 2018/11/21 13:32:48 (GMT -700)
  config file = None
  configured module search path = [u'/home/test/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/test/dev/ansible/lib/ansible
  executable location = /home/test/dev/ansible/bin/ansible
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by vexata** [comment]
@Akasurde  messed up branch for my previous PR #47091, so I've submitted a new PR with the changes. Tests will be submitted in later PRs.

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        )
+    )
+
+    module = AnsibleModule(arg_spec,
+                           supports_check_mode=True,
+                           required_together=required_together())
+
+    state = module.params['state']
+    size = module.params['size']
+    if size and size_to_MiB(size) < 0:
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py]: `delete_volume` does not require size. This check will force user to specify size even if he/she is deleting the volume. I feel this is not intuitive UI/UX.

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        module.fail_json(msg='Volume size should be in M, G, T units')
+
+    array = get_array(module)
+    volume = get_volume(module, array)
+
+    if state == 'present' and size and not volume:
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py]:
```python
if state == 'present':
    if not volume:
        create_volume(module, array)
    elif volume:
        update_volume(module, array, volume)
elif state == 'absent' and volume:
    delete_volume(module, array, volume)
else:
    module.exit_json(changed=False)
```

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            return vol[0]
+        else:
+            return None
+    except Exception:
+        module.fail_json(msg='Error while attempting to retrieve volumes.')
+
+
+def create_volume(module, array):
+    """"Create a new volume."""
+    changed = False
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py:107]:
```suggestion
    changed = False
    if not module.params.get('size', False):
        module.fail_json(msg='Size is required to update volume')
```

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 1 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py]: The assumption is incorrect, user is not forced to provide size for delete_volume by this code.

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 3 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py:107]: Refactored. Error message in review comment should be for create, not update.

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 2 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py]: Refactored

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 1 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py]: Refactor should make this non-dependence on size obvious

**Comment by vexata** [comment]
@Akasurde, please review. Have refactored the code and pushed the changes: [67b9cac](https://github.com/ansible/ansible/pull/49006/commits/67b9cac42e3b21fefc725a0e8ab6182fed0f86a4)

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
Are you planning to have unit tests for this change ?

**Comment by vexata** [comment]
@Akasurde, yes as I mentioned in an earlier reply, I have several more modules in the pipeline and unit tests for all will be checked in subsequent PRs

**Comment by Akasurde** [comment]
> @Akasurde, yes as I mentioned in an earlier reply, I have several more modules in the pipeline and unit tests for all will be checked in subsequent PRs

OK

**Comment by vexata** [comment]
@acozine do I need to do anything else for the merge to happen ? Thx.

File: lib/ansible/utils/module_docs_fragments/vexata.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/vexata.py]: Please rebase as this directory has moved to https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/doc_fragments/

File: lib/ansible/utils/module_docs_fragments/vexata.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+notes:
+    - Ansible modules are available for Vexata VX100 arrays.
+"""
+
+    # Documentation fragment for Vexata VX100 series
+    VX100 = '''
+options:
+  array:
+    description:
+      - Vexata VX100 array hostname or IPv4 Address.
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/vexata.py]: Please add `type:` for all options
```suggestion
      - Vexata VX100 array hostname or IPv4 Address.
    type: str
```

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    module.exit_json(changed=changed)
+
+
+def main():
+    arg_spec = argument_spec()
+    arg_spec.update(
+        dict(
+            name=dict(type='str', required=True),
+            state=dict(default='present', choices=['present', 'absent']),
+            size=dict()
```
Comment (gundalow) [inline lib/ansible/modules/storage/vexata/vexata_volume.py]: str?
```suggestion
            size=dict(type='str')
```

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        )
+    )
+
+    module = AnsibleModule(arg_spec,
+                           supports_check_mode=True,
+                           required_together=required_together())
```
Comment (gundalow) [inline lib/ansible/modules/storage/vexata/vexata_volume.py:183]: Is anything `required_together`, if not please remove this line

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 7 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py:183]: @gundalow, the username and password are required together

File: lib/ansible/utils/module_docs_fragments/vexata.py (hunk 5 above)
Comment (vexata) [reply lib/ansible/utils/module_docs_fragments/vexata.py]: Sure will do

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 6 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py]: Ok

File: lib/ansible/utils/module_docs_fragments/vexata.py (hunk 4 above)
Comment (vexata) [reply lib/ansible/utils/module_docs_fragments/vexata.py]: Will do.

**Comment by vexata** [comment]
@gundalow please let me know if this addresses everything:
https://github.com/ansible/ansible/pull/49006/commits/d05f25f505990421a47107258ac543f44a201175

**Comment by Akasurde** [comment]
Closing and re-opening for CI trigger.

**Review by gundalow — APPROVED** [review]

**Pull request merged**